### PR TITLE
Disha/hack feb/update tests genbank

### PR DESCRIPTION
**Pull request by Dishalodha**
Here I have created unit testing for extract_data from the genbank module
I have divided the tests into 2 files:

- test_extract_data.py : Tests all the methods that generate a output file
- test_extract_data_seq.py: Tests all the internal methods and formatting of the files

### CODE REVIEW

File: src/python/tests/genbank/test_extract_data.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+from pathlib import Path
+from unittest.mock import Mock, patch
+import pytest
+
+
+from BCBio import GFF
+from Bio import SeqIO
+from Bio.Seq import Seq
+from Bio.SeqRecord import SeqRecord
+from Bio.SeqFeature import SeqFeature, FeatureLocation
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data.py:31]:
```suggestion

from BCBio import GFF
from Bio import SeqIO
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.SeqFeature import SeqFeature, FeatureLocation
import pytest
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+# limitations under the License.
+"""Unit testing of `ensembl.io.genomio.genbank.extract_data` module.
+
+Typical usage example::
+    $ pytest test_extract_data_seq.py
+
+"""
+from unittest.mock import Mock, patch
+from typing import List
+from Bio.SeqRecord import SeqRecord
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:23]:
```suggestion
from typing import List
from unittest.mock import Mock, patch

from Bio.SeqRecord import SeqRecord
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+from Bio.Seq import Seq
+from Bio.SeqFeature import SeqFeature, FeatureLocation
+import pytest
+
+
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:28]:
```suggestion
```

**Review by JAlvarezJarreta — COMMENTED** [review]
_Partial review_

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        type_feature: str,
+    ):
+        """Test to parse the features correctly"""
+        record = SeqRecord(Seq("ATGC"), id="record1")
+        gene_feature = SeqFeature(FeatureLocation(10, 20), type="gene", qualifiers={gene_name: expected_name})
+        rna_feature = SeqFeature(FeatureLocation(10, 15), type=type_feature)
+        CDS_feature = SeqFeature(
+            FeatureLocation(10, 20), type="CDS", qualifiers={gene_name: "GlyrA", "transl_table": "2"}
+        )
+        record.features = [gene_feature, rna_feature, CDS_feature]
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:74]:
```suggestion
        cds_feature = SeqFeature(
            FeatureLocation(10, 20), type="CDS", qualifiers={gene_name: "GlyrA", "transl_table": "2"}
        )
        record.features = [gene_feature, rna_feature, cds_feature]
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        mock_peptides: List = []
+
+        assert gene_feature.qualifiers[gene_name] == expected_name
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:77]:
```suggestion
```
This is testing behaviour of biopython (that's on them to do, not us).

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        # pylint: disable=protected-access
+        formatted_files_generator._parse_record(record)
+        if gene_feature.qualifiers[gene_name]:
+            mock_parse_gene_feat.assert_called()
+
+        if rna_feature.type in ("tRNA", "rRNA"):
+            mock_parse_rna_feat.assert_called()
+
+    def test_parse_record_with_unsupported_feature(self, formatted_files_generator: FormattedFilesGenerator):
+        """Raises error then there is unspported feature"""
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:98]:
```suggestion
        """Tests parsing records with unsupported features."""
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            ("CDS", "AGR_01", "pseudo", "TESTAGR_01_p1"),
+        ],
+    )
+    def test_parse_gene_feat(
+        self,
+        expected_id: str,
+        gene_name: str,
+        type_feature: str,
+        test_qualifiers: str,
+        formatted_files_generator: FormattedFilesGenerator,
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:123]: Same as above.

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+            ("AGR90MT_t01_t2", "AGR90MT_t01_t2_t1", "AGR90MT_t01_t2"),
+        ],
+    )
+    @pytest.mark.dependency(name="rna_parse", depends=["parse_record"])
+    def test_parse_rna_feat(
+        self,
+        expected_gene_id: str,
+        expected_rna_id: str,
+        formatted_files_generator: FormattedFilesGenerator,
+        rna_name: str,
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:175]: Sames as before.

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    @pytest.mark.parametrize(
+        "gene_id, all_ids, expected_id",
+        [("gene_name", ["gene_name"], "gene_name_2"), ("gene_test", [""], "gene_test")],
+    )
+    def test_uniquify_id(
+        self,
+        all_ids: List[str],
+        expected_id: str,
+        formatted_files_generator: FormattedFilesGenerator,
+        gene_id: str,
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:198]: Same as before.

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    ) -> None:
+        """Test that _uniquify_id adds a version number to an existing ID"""
+        # pylint: disable=protected-access
+        new_id = formatted_files_generator._uniquify_id(gene_id, all_ids)
+        assert new_id == expected_id
+
+    @pytest.mark.parametrize("organelle, expected_location", [("mitochondrion", "mitochondrial_chromosome")])
+    def test_prepare_location_with_supported_organelle(
+        self, expected_location: str, formatted_files_generator, organelle: str
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:207]:
```suggestion
        self, formatted_files_generator: FormattedFilesGenerator, expected_location: str, organelle: str
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    @pytest.mark.parametrize("organelle", [("miton")])
+    def test_prepare_location_with_unsupported_organelle(
+        self, formatted_files_generator: FormattedFilesGenerator, organelle: str
+    ) -> None:
+        """Test that organelle location if not identifies throws an error"""
+        # An organelle not in the dictionary
+        with pytest.raises(UnsupportedData) as exc_info:
+            # pylint: disable=protected-access
+            formatted_files_generator._prepare_location(organelle)
+        assert str(exc_info.value) == f"Unknown organelle: {organelle}"
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:223]:
```suggestion
        with pytest.raises(UnsupportedData, match=f"Unknown organelle: {organelle}"):
            # pylint: disable=protected-access
            formatted_files_generator._prepare_location(organelle)
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+
+    @pytest.mark.parametrize(
+        "type_feature, expected_value", [("gene", None), ("mRNA", None), ("CDS", 2), ("CDS", 5)]
+    )
+    def test_get_codon_table(
+        self, expected_value: str, formatted_files_generator: FormattedFilesGenerator, type_feature: str
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:229]: Same as before.

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
A few things to review to make the code more homogeneous. On top of the comments made, a couple of general ones:
* It would be nice if `@patch`, `@dependencies` and `@parametrize` where all in the same order in all the tests (that is not the case now)
* I believe `formatted_files_generator` fixture can be move into a `conftest.py` file located in `src/python/tests/genome_metadata`, so every test inside can access it. Alternatively, you can create the object by default once for every test [example](https://github.com/Ensembl/ensembl-compara/blob/release/111/src/python/tests/test_repair_mlss_tags.py#L46-L56) and take advantage of having all tests within a Test class.

File: src/python/ensembl/io/genomio/genbank/extract_data.py [hunk 13]
```diff
@@ -523,7 +523,7 @@ def _write_genome_json(self, genome_data: Dict[str, Any]):
 
 def main() -> None:
     """Main script entry-point."""
-    parser = ArgumentParser(description="Parse a GenBank file and create cleaned up files from it.")
+    parser = ArgumentParser(description=__doc__)
```
[the remark this answers is not in the document]
Comment (Dishalodha) [reply src/python/ensembl/io/genomio/genbank/extract_data.py:521]: okay yes sure

File: src/python/ensembl/io/genomio/genbank/extract_data.py [hunk 14]
```diff
@@ -518,7 +518,7 @@ def _write_genome_json(self, genome_data: Dict[str, Any]):
 
 def main() -> None:
     """Main script entry-point."""
-    parser = ArgumentParser(description=__doc__)
+    parser = ArgumentParser(description="Parse a GenBank file and create cleaned up files from it.")
```
[the remark this answers is not in the document]
Comment (Dishalodha) [reply src/python/ensembl/io/genomio/genbank/extract_data.py:521]: I am not sure, the module's docstring is a complete explaination of what we expect in this module and maybe its better to keep that?

**Review by JAlvarezJarreta — APPROVED** [review]
I'm happy with the current status of this code :+1:

File: src/python/tests/genbank/test_extract_data.py [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+from BCBio import GFF
+from Bio import SeqIO
+from Bio.Seq import Seq
+from Bio.SeqFeature import SeqFeature, FeatureLocation
+from Bio.SeqRecord import SeqRecord
+import pytest
+
+from ensembl.io.genomio.genbank.extract_data import FormattedFilesGenerator, GBParseError
+from ensembl.io.genomio.utils import get_json
+from pytest import TempPathFactory
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data.py:35]:
```suggestion
```

File: src/python/tests/genbank/test_extract_data.py [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+from pathlib import Path
+from typing import List
+from unittest.mock import Mock, patch
+
+from BCBio import GFF
+from Bio import SeqIO
+from Bio.Seq import Seq
+from Bio.SeqFeature import SeqFeature, FeatureLocation
+from Bio.SeqRecord import SeqRecord
+import pytest
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data.py:31]:
```suggestion
import pytest
from pytest import TempPathFactory
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+from typing import List
+from unittest.mock import Mock, patch
+
+from Bio.Seq import Seq
+from Bio.SeqFeature import SeqFeature, FeatureLocation
+from Bio.SeqRecord import SeqRecord
+import pytest
+
+from ensembl.io.genomio.genbank.extract_data import FormattedFilesGenerator, GBParseError, UnsupportedData
+from pytest import TempPathFactory
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:32]:
```suggestion
```

File: src/python/tests/genbank/test_extract_data_seq.py [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+"""
+
+from pathlib import Path
+from typing import List
+from unittest.mock import Mock, patch
+
+from Bio.Seq import Seq
+from Bio.SeqFeature import SeqFeature, FeatureLocation
+from Bio.SeqRecord import SeqRecord
+import pytest
```
Comment (JAlvarezJarreta) [inline src/python/tests/genbank/test_extract_data_seq.py:29]:
```suggestion
import pytest
from pytest import TempPathFactory
```

**Review by JAlvarezJarreta — APPROVED** [review]
Couple of minor changes, but the code is good to go!

**Pull request merged**